CSC8503/InventoryBuffSystem/PickupGameObject.cpp

```diff
@@ -0,0 +1,130 @@
+#include "PickupGameObject.h"
+#include "StateTransition.h"
+#include "StateMachine.h"
+#include "State.h"
+#include "PhysicsObject.h"
+#include "Vector3.h"
+#include "random"
+#include "InventoryBuffSystem.h"
+
+using namespace NCL;
+using namespace CSC8503;
+using namespace InventoryBuffSystem;
+
+PickupGameObject::PickupGameObject(std::map<GameObject*, int>* mPlayerObjectToPlayerNoMap, float initCooldown) {
+	mCooldown = 0.0f;
+	mInitCooldown = initCooldown;
+	mPlayerObjectToPlayerNoMap = mPlayerObjectToPlayerNoMap;
+
+	mStateMachine = new StateMachine();
+	State* WaitingState = new State([&](float dt) -> void
+		{
+			this->Waiting(dt);
+		}
+	);
+	State* GoUnderState = new State([&](float dt) -> void
+		{
+			this->GoUnder(dt);
+		}
+	);
+	State* GoOverState = new State([&](float dt) -> void
+		{
+			this->GoOver(dt);
+		}
+	);
+
+	mStateMachine->AddState(WaitingState);
+	mStateMachine->AddState(GoUnderState);
+	mStateMachine->AddState(GoOverState);
+
+	mStateMachine->AddTransition(new StateTransition(WaitingState, GoUnderState,
+		[&]() -> bool
+		{
+			return this->mCooldown == INT_MAX;
+		}
+	));
+
+	mStateMachine->AddTransition(new StateTransition(GoUnderState, WaitingState,
+		[&]() -> bool
+		{
+			return this->mCooldown <= this->mInitCooldown;
+		}
+	));
+
+	//InitCooldown -1 indicates a pickup that will not "Respawn"
+	//So other transitions will not be necessary
+	if (mInitCooldown == -1)
+		return;
+
+	mStateMachine->AddTransition(new StateTransition(WaitingState, GoOverState,
+		[&]() -> bool
+		{
+			return this->mCooldown <= 0.2f && this->mCooldown > 0;
+		}
+	));
+
+	mStateMachine->AddTransition(new StateTransition(GoOverState, WaitingState,
+		[&]() -> bool
+		{
+			return this->mCooldown == 0;
+		}
+	));
```

Comment:
The brackets on these functions should adhere to the code standards
mStateMachine->AddTransition() {

}